Date: Oct 01 09:33:25 2018

From: d3767f0cd38473f296f66d157495fc9b655ccfa2 Dec 26 16:11:01 2018

Subject: [PATCH] XRE-13615 : Update to latest metro revision in 2017 branch

	Patch to add Cookie APIs in libsoup-2.4 to avoid compilation error in latest revision on wpe-20170728

Source: COMCAST

License: LGPL-2.0-or-later

Upstream-Status: Pending

Signed-off-by: mgopal003c <Manigandan_Gopalakirshnan@cable.comcast.com>
---

--- libsoup-2.58.1.orig/libsoup/soup-cookie-jar.c	2018-07-11 10:09:41.062845111 +0000
+++ libsoup-2.58.1/libsoup/soup-cookie-jar.c	2018-07-11 10:46:56.708218997 +0000
@@ -49,6 +49,8 @@
 	GHashTable *domains, *serials;
 	guint serial;
 	SoupCookieJarAcceptPolicy accept_policy;
+	guint64 n_cookies;
+	guint64 limit;
 } SoupCookieJarPrivate;
 
 static void soup_cookie_jar_session_feature_init (SoupSessionFeatureInterface *feature_interface, gpointer interface_data);
@@ -58,6 +60,8 @@
 			 G_IMPLEMENT_INTERFACE (SOUP_TYPE_SESSION_FEATURE,
 						soup_cookie_jar_session_feature_init))
 
+#define MAX_COOKIE_SIZE 102400 /* 100K */
+
 static void
 soup_cookie_jar_init (SoupCookieJar *jar)
 {
@@ -233,6 +237,18 @@
 	return g_object_new (SOUP_TYPE_COOKIE_JAR, NULL);
 }
 
+void
+soup_cookie_jar_set_limit (SoupCookieJar *jar,
+               guint64        limit)
+{
+	SoupCookieJarPrivate *priv;
+
+	g_return_if_fail (SOUP_IS_COOKIE_JAR (jar));
+
+	priv = soup_cookie_jar_get_instance_private (jar);
+	priv->limit = limit;
+}
+
 /**
  * soup_cookie_jar_save:
  * @jar: a #SoupCookieJar
@@ -257,10 +273,13 @@
 {
 	SoupCookieJarPrivate *priv = soup_cookie_jar_get_instance_private (jar);
 
-	if (old && old != new)
+	if (old && old != new) {
 		g_hash_table_remove (priv->serials, old);
+		priv->n_cookies--;
+	}
 	if (new) {
 		priv->serial++;
+		priv->n_cookies++;
 		g_hash_table_insert (priv->serials, new, GUINT_TO_POINTER (priv->serial));
 	}
 
@@ -432,6 +451,15 @@
 	return get_cookies (jar, uri, for_http, TRUE);
 }
 
+static unsigned cookie_size (SoupCookie *cookie)
+{
+	/* Consider name, value, domain and path only for the size */
+	return (cookie->name ? strlen(cookie->name) : 0) +
+		(cookie->value ? strlen(cookie->value) : 0) +
+		(cookie->domain ? strlen(cookie->domain) : 0) +
+		(cookie->path ? strlen(cookie->path) : 0);
+}
+
 /**
  * soup_cookie_jar_add_cookie:
  * @jar: a #SoupCookieJar
@@ -462,6 +490,11 @@
 		return;
 	}
 
+	if (cookie_size (cookie) > MAX_COOKIE_SIZE) {
+		g_warning ("Trying to add a cookie for domain '%s', bigger than maximum allowed size %d, ignoring it", cookie->domain, MAX_COOKIE_SIZE);
+		return;
+	}
+
 	priv = soup_cookie_jar_get_instance_private (jar);
 	old_cookies = g_hash_table_lookup (priv->domains, cookie->domain);
 	for (oc = old_cookies; oc; oc = oc->next) {
@@ -497,6 +530,12 @@
 		soup_cookie_free (cookie);
 		return;
 	}
+
+	if (priv->limit > 0 && priv->n_cookies + 1 >= priv->limit) {
+		g_warning ("Reached limit of %" G_GUINT64_FORMAT " cookies, not adding the new one for domain '%s'", priv->limit, cookie->domain);
+		soup_cookie_free (cookie);
+		return;
+	}
 
 	if (last)
 		last->next = g_slist_append (NULL, cookie);
--- libsoup-2.58.1.orig/libsoup/soup-cookie-jar.h	2018-07-11 10:09:41.054845164 +0000
+++ libsoup-2.58.1/libsoup/soup-cookie-jar.h	2018-07-11 10:19:23.071032906 +0000
@@ -52,6 +52,9 @@
 SOUP_AVAILABLE_IN_2_24
 SoupCookieJar *           soup_cookie_jar_new                         (void);
 SOUP_AVAILABLE_IN_2_24
+void                      soup_cookie_jar_set_limit                   (SoupCookieJar             *jar,
+								       guint64                    limit);
+SOUP_AVAILABLE_IN_2_24
 char          *           soup_cookie_jar_get_cookies                 (SoupCookieJar             *jar,
 								       SoupURI                   *uri,
 								       gboolean                   for_http);
